Enable mypy for ngclient

sechkova opened the pull request:
Fixes #1409 
**Description of the changes being introduced by the pull request**:

Enable static type checking for ngclient.
- add all missing type annotations
- fix errors (partially for updater.py and trusted_metadata_set.py
- add urllib3 to ignored includes
- download suggested stub packages when running mypy

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
Add ngclient to mypy

Extend mypy to include all files under ngclient.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add type annotations to download.py and fetcher.py

Add the stub for the requests package (types-requests)
to requirements-tests.txt.

Add urllib3 to the ignored imports. The project seems
to have added type annotations already but has not
released a version including them yet.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Replace BinaryIO with IO[bytes] in metadata.py

Needed in order to be compatible with the return type of
download_file (TemporaryFile is typed as IO[bytes]).
BinaryIO is a subclass of IO[bytes].

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add missing type annotations to updater

Add missing annotations and partially resolve mypy
errors in updater.py and trusted_metadata_set.py

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku commented:
I think there's a few cases here:
* TrustedMetadataSet.root probably does not actually need to be Optional: Unlike other properties it's guaranteed to be a Metadata after constructor, right?
* mypy gets confused about target_base_url: I'm sure it's possible to simplify the code so mypy understands that target_base_url cannot be None
* cases where we know a e.g. self._trusted_set.snapshot is Metadata but code does not: I think this is a great place for `assert(self._trusted_set.snapshot is not None)` 
* many cases of Metadata.Signed typing problem...
  * I believe this would get fixed with #1457 and Generics
  * alternatively we could just use calls like `assert(isinstance(self._trusted_set.snapshot.signed, Snapshot))`. Ugly but works and there are surprisingly few places where this is needed (like 10-12 or so?)
  * even more alternatively we could add properties to TrustedMetadataSet that returned Root, Snapshot, Timestamp and Targets (but that makes a lot of the code a bit ugly as you usually need both the Metadata and the Signed subclass)...

First two-three can be fixed now. The last one we can discuss... In any case this is useful work as it shows how much (or little) the lack of signed typing affects us...

The download related the changes should maybe only be done after the cleanup is done in #1448 (also the SlowRetrievalError change looks a bit fishy -- I would expect it's used in the legacy code as well?)

joshuagl commented:
> * mypy gets confused about target_base_url: I'm sure it's possible to simplify the code so mypy understands that target_base_url cannot be None

It's debatable whether this simplifies things for humans reading the code (edit: though, apart from the variable name ugliness and possibility for confusion of `tgt_base_url`, I think it probably is simpler) but I was able to make mypy happy with the following:

```Python
        # locally scoped variable that is definitely a string, even if empty
        tgt_base_url: str = self._target_base_url or ""
        if target_base_url is not None:
            tgt_base_url = _ensure_trailing_slash(target_base_url)

        # move the check for a base URL being set to after the locally scoped
        # variable assignment logic is bottomed out.
        if not tgt_base_url:
            raise ValueError(
                "target_base_url must be set in either download_target() or "
                "constructor"
            )
```

see the [full diff here](https://github.com/joshuagl/tuf/commit/2aaec55ab303bf9b753c863df835d9be50c4a2bd).

jku commented:
this should also work -- the issue for mypy is two branches that seem unrelated (even if they aren't) so making the branches actually same should make the intent clear to computer...
```
        if target_base_url is None:
            if self._target_base_url is None:
                raise ValueError(
                    "target_base_url must be set in either download_target() or "
                    "constructor"
                )
            target_base_url = self._target_base_url
        else:
            target_base_url = _ensure_trailing_slash(target_base_url)
```

jku commented:
> The download related the changes should maybe only be done after the cleanup is done in #1448 (also the SlowRetrievalError change looks a bit fishy -- I would expect it's used in the legacy code as well?)

#1448 is now merged

sechkova committed:
Make TrustedMetadataSet.root non-optional

The 'root' property is guaranteed to be set after init.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Rebased on latest changes in develop and fixed all issues but the Metadata.Signed typing ones.

sechkova commented:
Actually `bandit` complains slightly about asserts:
```
>> Issue: [B101:assert_used] Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.
   Severity: Low   Confidence: High
   Location: tuf/ngclient/updater.py:297
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html
296	
297	            assert self._trusted_set.snapshot is not None
298	            metainfo = self._trusted_set.snapshot.signed.meta[f"{role}.json"]
```

joshuagl commented:
> Actually `bandit` complains slightly about asserts

Right. We could reasonably silence those warnings, as we are only suggesting use of asserts to help mypy determine the expected type. But I think this further reinforces the 'code smell' of using asserts to solve this problem.

sechkova commented:
> > Actually `bandit` complains slightly about asserts
> 
> Right. We could reasonably silence those warnings, as we are only suggesting use of asserts to help mypy determine the expected type. But I think this further reinforces the 'code smell' of using asserts to solve this problem.

In this case we use assert as the [recommended way](https://mypy.readthedocs.io/en/stable/common_issues.html#unexpected-errors-about-none-and-or-optional-types) to deal with `None` errors. So we either silence bandit or tweak the TrustedMetadataSet public properties to not return optional values.

sechkova committed:
Denote expected type of Metadata.signed

By explicitly denoting the expected type of Metadata.signed
we help mypy understand our intentions and correctly figure
out types. This is entirely a typing feature and has no
runtime effect.

Modify the return type of Metadata.from_dict to match the
other factory methods (from_*).

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Use assert to check for None values

This is done only for hinting 'mypy' that we have
ensured these values cannot be None.

'Bandit' raises warnings for assert usage in the code
but we are disabling them since we do not rely upon
'asserts' for any runtime checks but only for type checking.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
I used the part of #1457 that makes Metadata 'Generic' to resolve the type checking issues related to the 'signed' type.
Now there are no more mypy errors left and the PR should be ready for review.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -730,7 +730,7 @@ class BaseFile:
     @staticmethod
     def _verify_hashes(
-        data: Union[bytes, BinaryIO], expected_hashes: Dict[str, str]
+        data: Union[bytes, IO[bytes]], expected_hashes: Dict[str, str]
```
`BinaryIO` appears to be a subclass of `IO[bytes]` with the addition of interfaces to support context managers (`with`). We are not actually using the passed `data` object as a context manager, so this looks fine.

joshuagl commented on tuf/ngclient/updater.py:
```diff
@@ -111,12 +112,7 @@ def __init__(
-        else:
-            self._fetcher = fetcher
-
+        self._fetcher = fetcher or requests_fetcher.RequestsFetcher()
```
Nice 👍

joshuagl commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -152,7 +152,7 @@ def __len__(self) -> int:
     def __iter__(self) -> Iterator[Metadata]:
         """Returns iterator over all Metadata objects in TrustedMetadataSet"""
-        return iter(self._trusted_set)
+        return iter(self._trusted_set.values())
```
Nice catch! thanks to mypy?

sechkova replied:
Yes, mypy gets the credit.

joshuagl commented on tuf/ngclient/updater.py:
```diff
@@ -143,7 +139,7 @@ def refresh(self) -> None:
         self._load_targets("targets", "root")
 
-    def get_one_valid_targetinfo(self, target_path: str) -> Dict:
+    def get_one_valid_targetinfo(self, target_path: str) -> Dict[str, Any]:
```
I wonder if there is a better return type we could be using here? Rather than a dict which will only ever contain two key-value pairs where the values have different types?

I am not suggesting we change something here in this PR, just thinking out loud.

sechkova replied:
Yeah, I think there is a common "willingness" and a bunch of issues related to improving this part of the API.
The suggestion here seems to be the closest to yours: https://github.com/theupdateframework/tuf/issues/1411

joshuagl commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -205,7 +205,7 @@ def update_root(self, data: bytes) -> None:
             )
 
-        if self.root is not None:
+        if self._trusted_set.get("root") is not None:
```
If the root property is guaranteed to be set after init, do we need this check?

sechkova replied:
We need it because we call `update_root` during init too ... to validate the initial root.
Do you think we need to clarify it somehow?

sechkova replied:
Since you expressed some worries about inconsistently using the 'root' property, the other solution I came up with is this  https://github.com/theupdateframework/tuf/pull/1489/commits/8c172af244c968b221bcbd5398a7a07a6c897408

jku replied:
this looks good to me at least

joshuagl commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -196,7 +196,7 @@ def update_root(self, data: bytes) -> None:
 
         try:
-            new_root = Metadata[Root].from_bytes(data)
+            new_root: Metadata[Root] = Metadata[Root].from_bytes(data)
```
These type hints for the IDE are starting to look quite cumbersome. Can the IDE not recognise the type hint on the right-hand side of the assignment operator? Do we really need them?😕

sechkova replied:
Now that you made me have another look :roll_eyes: .... it turned out I wasn't able to annotate the return type of the `from_*` methods with the generic type 'T' because `from_bytes` was a static method. So I amended the last commit and I think this one now looks better and does the job: 6754470

joshuagl reviewed (commented):
Thanks Teodora. This is looking good. I have a minor comment on the change to the root property, and a slight pushback on the type hinting for IDE's that I would appreciate your thoughts on before we merge.

sechkova commented:
Seems like the CI won't be triggered until I rebase and resolve the conflicts.

jku commented:
> * add urllib3 to ignored includes

If you researched this can you document why? looking in urllib3 git the annotations seem to be there but is that just for some future 2.0 release?

> * download suggested stub packages when running mypy

Could we add the required stub packages to test requirements instead?

jku commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -97,7 +97,7 @@ def __init__(self, root_data: bytes):
                 error type and content will contain more details.
         """
-        self._trusted_set = {}  # type: Dict[str: Metadata]
+        self._trusted_set = {}  # type: Dict[str, Metadata]
```
```suggestion
        self._trusted_set: Dict[str, Metadata] = {}
```
Since we're fixing my broken annotations, let's do it like everywhere else :grimacing:

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -141,7 +141,7 @@ def _chunks(
 
         except urllib3.exceptions.ReadTimeoutError as e:
-            raise exceptions.SlowRetrievalError(str(e))
+            raise exceptions.DownloadError from e
```
Reasons for this are good but now the documented exceptions (here and in Fetcher API) are incorrect... Hmm. Maybe you could make SlowRetrievals speed argument optional instead?

jku replied:
or figure out another solution -- I didn't spend a lot of time on that suggestion

sechkova replied:
Yep, optional seems to fit well: 7cbfd0e

jku reviewed (commented):
I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving). 

This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already

The last commit is IMO not great since it has so much code in the try-block: it's not clear that the KeyError should only happen in `self.root`... I think the original code is better. Maybe the real alternative would be to not call `update_root()` from init but instead do the required steps in init itself -- that seems ok to me as well as I think there aren't that many steps here.

sechkova commented:
> I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving).

Hmm ... I wouldn't say directly depending, I didn't cherry-pick the commit from the other PR but created a new one and added you as a co-author.
 
> This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already

This was done intentionally but I guess I should've tried to explain my intentions somewhere in written form. 

The "Generics" branch tries to achieve two things simultaneously: correct type annotations of Signed and a runtime check ensuring  content and type are really what we expect. Since there seems to be an agreement already on the use of Generics and an ongoing discussion around the runtime check part, I decided to include only the first part here, the one directly related to type annotations and mypy. This way we can leave the rest of the discussion for #1457 which will also make it simpler because it will try to solve only one problem. I hope you agree :)

sechkova commented:
> The last commit is IMO not great since it has so much code in the try-block: it's not clear that the KeyError should only happen in `self.root`... I think the original code is better. Maybe the real alternative would be to not call `update_root()` from init but instead do the required steps in init itself -- that seems ok to me as well as I think there aren't that many steps here.

I tried this alternative since @joshuagl expressed some concerns about not using the 'root' property consistently inside the class. The three options, none of them perfect, would be:
- don't call `update_root` from `init`, repeat some of the code (not than much repetition so could be acceptable)
- don't use the property but "get" safely from `self._trusted_set`  c205ac0 (concerns about not using consistently the root property)
- use the root property, enclose it in a try bock and catch the KeyError 8c172af (well you already described the disadvantages)

There is a forth one where we keep root optional as the rest of the properties and add asserts in the code but I tried and this means asserts at 4-5 places which seems to rule out this option.

jku commented:
> > I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving).
> 
> Hmm ... I wouldn't say directly depending, I didn't cherry-pick the commit from the other PR but created a new one and added you as a co-author.
> 
> > This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already
> 
> This was done intentionally but I guess I should've tried to explain my intentions somewhere in written form.

Yeah I think the generics change is major enough that we can't just hide it in a PR titled "enable mypy for ngclient" :) We can definitely split the PR if that makes it easier to swallow though.

joshuagl commented:
> Yeah I think the generics change is major enough that we can't just hide it in a PR titled "enable mypy for ngclient" :) We can definitely split the PR if that makes it easier to swallow though.

Ack, let's do it.

sechkova commented:
> > * add urllib3 to ignored includes
> 
> If you researched this can you document why? looking in urllib3 git the annotations seem to be there but is that just for some future 2.0 release?

Yes, the latest release is 1.26 which does not include the type annotations. I updated the commit description.

> > * download suggested stub packages when running mypy

> Could we add the required stub packages to test requirements instead?

Done:  d306370

sechkova committed:
Make SlowRetrievalError parameter optional

SlowRetrievalError is raised from RequestsFetcher where
average_download_speed is not calculated.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku commented:
Ok, I've made the #1457 PR into a "minimal generics" PR -- this PR could be based off of that one I believe.

sechkova committed:
Load trusted root in a separate private method

Add an additional private method for loading the initial
trusted root metadata. The public method update_root() is
now used only externally for updating the intiial root.
The 'root' property is used only after its initialization
in the constructor and is not longer optional which makes
mypy happy.

This split results in cleaner code and the ability to annotate
the 'root' property as non-optional at the cost of some code
duplication.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Rebased on #1457 and I believe I have addressed all comments so waiting for another review.

jku commented on tuf/ngclient/updater.py:
```diff
@@ -348,6 +345,7 @@ def _load_snapshot(self) -> None:
             # Local snapshot does not exist or is invalid: update from remote
             logger.debug("Failed to load local snapshot %s", e)
 
+            assert self._trusted_set.timestamp is not None  # nosec
```
this is unfortunate... I don't have any better ideas but it looks so strange in this code.

This seems to only be needed in Updater for getting meta-info -- which is also a sort of duplicate piece of code here and in `_load_targets()`... so there's probably some sort of refactoring possibility here but let's not do that in this PR.

jku commented on tuf/api/metadata.py:
```diff
@@ -98,7 +98,7 @@ def __init__(self, signed: T, signatures: "OrderedDict[str, Signature]"):
 
     @classmethod
-    def from_dict(cls, metadata: Dict[str, Any]) -> "Metadata":
+    def from_dict(cls, metadata: Dict[str, Any]) -> "Metadata[T]":
```
I don't object but I'm curious did this actually affect something? I expect from_dict() to be only called from Deserializer which does not know about specific signed types.

sechkova replied:
No, I did it just for consistency with the others.

jku reviewed (approved):
Looks good, updater.py in VSCode is much nicer now. I think this is good to go... except this will definitely clash with #1514 which I was about to merge...

sechkova commented:
Rebased on latest changes in develop.
